WIP RFC smoothly switch crop on/off

dterrahe opened the pull request:
Another follow up to #15084 and #15364. Fixes #15416.

The visible effect of this PR is twofold:
- when the crop of the image in darkroom is switched on or off, for example because the crop module itself or the retouch or liquify modules are (de)focused, the previously visible high resolution image is repositioned and rescaled to the exact new spot where it will end up, while the full image is being recalculated. When "zooming in" (the crop module loses focus; the crop is switched on) this means you should immediately see the correct image, but after the recalculation finishes it will be a little sharper. When "zooming out" (the crop/retouch/liquifty module is focused; the crop is temporarily switched off) this means that the "center" of the image will likely be sharper for a little while than the area around it, which may first be black and then pixilated while a scaled up version of the left-top corner preview is shown until the full recalculation finishes.
In the most extreme case (the crop margins are zero) this should avoid the problem @TurboGit reported here #15286 (no flickering), but it should help in a lot more cases. It also replaces the hacky solution I implemented here #15222.
- when we are currently zoomed in to a cropped image, focusing a module that temporarily switches off cropping (like crop itself but also retouch and liquify) would cause the shown image to move to a completely different area. This is no longer the case, although the image may still be slightly repositioned to move it to center.



This PR was a hellish journey again. To help understand what was needed to be changed and possibly predict/understand some side effects, I'll give a bit of background, also serving maybe as some kind of "documentation" when further changes are considered. Such documentation is completely lacking at the moment as far as I can tell and otherwise probably would not really reflect the actual current implementation and behavior anyway, so might have been worse than the nothing we have.

In the darkroom we have a multitude of coordinate systems to describe a spot in the image:
1. relative position in terms of the "complete" image (the one shown in miniature in the top-left navigator window. This goes from horizontal 0-100% (left-right) and vertical 0-100% (top-bottom).
2. same, but in pixels 0-720 horizontal 0-350 vertical (or a portion of this based on aspect ratio).
3. actually, the buffer defined in 2. is only partially used if `preview_downsampling` is set to 1/2, 1/3 or 1/4.
4. pixel position within the full calculated image based on applied crops/framing/rotation etc. This could be, say 0-3k horizontal and 0-2k vertical for a 24mpixel image that is cropped 50% in both directions.
5. position within the central area of the darkroom. On an FHD monitor, full screen and with all panels hidden, this would be 0-1920 and 0-1080. This still includes a narrow border around the image (set in prefs) and a big white+black border if color calibration mode is switched on.
6. relative position within the currently shown (zoomed) image in the central area of the darkroom (0-100% both).
7. same, but in pixels. This is 5 minus the borders.
8. - 10. same as 5-7 but for the second preview window.
11. corresponding position of any of the previous points in terms of the original raw file. In 0-100%. The 50% crop in 4. could for example be 50-100% and 25-75%, if the right half of the image positioned centrally vertically was used.
12. same in pixels; 3k-6k and 1k-3k for the example above.

These coordinate system are used while redrawing the screen and processing mouse moves and clicks and when dealing with visual elements like mask shapes, color picker areas and crop or perspective handles etc etc. Variables names don't indicate what coordinate system an "x" or "y" lives in (or at least I haven't discovered the consistency and I haven't myself improved the situation I'm sure). In _most_ places, we somehow reference back to a position within the "full" (i.e. current crop) image, but scaled down, that is shown in the navigator window (2. i.e. not necessarily corrected for downsampling; this correction needed to be applied every time before anything was done with it, so _all_ the individual modules etc were dealing with downsampling over and over again).

We have "current positions" for the center screen and the second preview window and the last calculated zoomed in "preview" for each of these, which may have been from a time that a crop was or wasn't active that currently isn't or is.

Basically, this PR stores the current position and the positions of the calculated previews in coordinate system 12 (the context of the RAW file, which is immutable) and then when drawing positions them on the center and second preview areas based on current crop and zoom levels and locations.

That sounds trivial and to avoid introducing a lot of new bugs one obviously tries to be minimally invasive when making changes and reconnecting to existing code. But similar past changes introduced back and forth conversions everywhere to the extend that it really impossible to comprehend what is going on. So I had to try to create some order out of chaos. Again, starting with not making too big changes so that the impact could more easily be tested, but eventually in frustration, because it was impossible to get everything to line up, ending up completely ripping out _all_ the references to downsampling and instead at startup reducing the dimensions of the "preview" (navigator) miniature. This miraculously got rid of a lot of issues with no problems introduced as far as I could tell (please _do_ tell if you find any). Similarly, after calculating a preview image (full res) for the zoomed+cropped, the resoluting pixmap would be stored in a variable (with associated dimension variables) called "backbuf". For some reason, at some point a duplicate of these variables called "output_backbuf" was created and extensively used. Because the dimensions of the pixmap are somewhat related (after scaling) to the common "standard" of the full "preview" pixmap, the code ended up with references to `output_backbuf_width/height` everywhere.

The output buffers are only available once the calculations are finished (which doesn't happen at the same time for the preview/full/preview2 pipes) so their sizes and zoom levels get out of sync with the current zoom and crop. This PR tries to base the shown image on the actual current zoom and crop levels, but unfortunately can't convert RAW dimensions back and forth to the current zoomed-and-cropped screen image until a pipe has been refreshed based on the current history (which gets partly flushed whenever a module (un)focus happens) and might not get recreated soon if a long calculation is already underway. And many modules don't draw their visual elements anyway until the _preview_ is available...

So, yeah, this PR is not perfect. Further improvements will need work done on cache and possibly pipe handling. I may start an issue to discuss my thoughts there first, because I definitely want to tap into the knowledge of @TurboGit and @jenshannoschwalm  and possibly others who currently know more about those areas than I do, because the changes required would be somewhat extensive and gradual steps are not likely to be productive.

Things you will notice when testing this PR:
- snapshots and duplicates work, but if they are activated while the currently visible image has a wildly different size, crop or  orientation, they will be shown in an odd location.
- slideshow is not connected up to the new approach at all yet; it doesn't work.
- When rotating the image, the scaled preview and repositioned last calculated full view will be temporarily combined in an odd juxtaposition. It is probably necessary to determine and store the rotation (and flip) of a calculated image relative to the original raw when a pipe calculation finishes and the result is stored in the backbuf. We can then ask cairo to rotate+flip as an approximation or just temporarily hide the last full res image. Previously this wasn't an issue because we just continued showing the old image in the same location while the recalc was ongoing; we were not attempting to relocate/zoom to a predicted new location.

Please give this PR a go and tell me your thoughts; any specific cases that you feel might need special handling and are not getting it here. Any showstoppers that might make it practically infeasible to get this approach working correctly ever. Or just any questions the above has raised for you or that could be explained (much) better.

jenshannoschwalm commented:
That's a bumpy and dirty road your on 😎

I just read through the diffs and so far wouldn't think there could be any problems interfering with iop cache.

We might have small glitches with roi aligning but will spot them pretty early  via -d pipe.

Great stuff principally, will test and check in detail the next days.

pehar1 commented:
A first quick report after some testing with a completely fresh install of your commit, image used for testing same as in issue 15416, with "reduced" xmp :

- issue 15416 seems fixed. Even after extensive CTRL-scrolling and panning I was not able to reproduce the issues, neither the black blocks, nor the mispositioning when the mouse is off-center during CTRL-scrolling.

> Crop "zooming in" ... but after the recalculation finishes it will be a little sharper.

> Crop "zooming out" ... and then pixilated while a scaled up version of the left-top corner preview is shown until the full recalculation finishes.

This is recognizable (with a "light" history stack)  only if OpenCL is disabled. With OpenCL enabled the update is too fast on my system to be visible with a "light" history stack. Adding a really "heavy" _Diffuse or Sharpen_ makes the effect visible but for me this is an expected behavior during a recalculation.

> ... when we are currently zoomed in to a cropped image ... image may still be slightly repositioned to move it to center.


This repositioning is visible but absolutely acceptable in my opinion.

dterrahe commented:
> wouldn't think there could be any problems interfering with iop cache.

So far this doesn't affect interaction with the iop cache at all. I'll see if today I can write a start to a discussion issue that gives an impression of what kind of improvements I think this PR would theoratically enable in cache usage.

> small glitches with roi aligning but will spot them pretty early via -d pipe.

Can you give a brief idea what kind of thing you think I should be looking out for?

Many thanks for your initial response and interest!

TurboGit commented:
> This PR was a hellish journey again.

I can easily understand.

> * This is no longer the case, although the image may still be slightly repositioned to move it to center.

I can see that, but very slightly and far better that what we have today for sure. And testing now with another image I have no repositioning visible at all. So it seems to depend on the actual crop as I have made my tests at 100% in both cases. 

One thing I can experience is a big "flash" on screen when opening Retouch or Liquify and there is a crop in the history.

Well I cannot reproduce now with OpenCL on my work machine. Yet if the image is cropped we see the crop area displayed on screen (at the wrong place) and the it disappear. To reproduce:

- Take a portrait picture
- Do a square crop at the center (that is activate module, select crop / square, leave module)
- Now zoom at 100%
- Expand/Collapse the Retouch or Liquify module

You should see the crop lines flashing.

TurboGit commented:
@dterrahe : First of all thanks for taking the lead on such hard issue. After some testing it seems to me that you are definitely on the right track as I have not yet found a big issue. I've not read the code yet.

What I have tested so far:

- Liquify with and without crop
- Retouch with and without crop
- With & without masking (drawn & parametric) on different modules

- Tested the duplicate module
- Tested the snapshot module - a slight issue when moving the separator and dragging it seems there is a lag. As far as I remember we had nothing while dragging so maybe now there is a refresh of the displayed area which is not needed maybe?

I have seen some skull which seems to indicate some refresh issue. One way to reproduce is to create a duplicate from the duplicate manager.

dterrahe commented:
> with another image I have no repositioning visible at all. So it seems to depend on the actual crop

If an image doesn't fill the whole center view in either direction it will always be centered. So if cropped and/or uncropped there is some area around it and it is a different width between the two and between the different sides it will _have_ to move. This is not a "bug" that can be fixed. _Unless_ we want to introduce an "unbounded" mode that allows the image to be positioned off-center. I had the intention to do this (so that shapes that are partly outside the image border can be more easily manipulated; holding ctrl while dragging could allow the image to move without snapping it to the window edge) but unless done carefully and intuitively this could easily lead to user complaints or to bugs of the kind shown here #15416.

> You should see the crop lines flashing.

I think I see what you mean. This may have something to do with the race/synchronisation issue that I allude to in the OP (paragraph starting with "The output buffers are only available [...]". The problem is that when switching the viewport crop there may be a period when the pipe that is used to transform back and forth between image and RAW dimensions has already been recreated but the size of the image has not yet been updated. Or something like that. I need to look at that again with a fresh mind at some point.

And then I'll try some "final" cleanup to make sure we have consistency. Previously we always referred to the (scaled down) preview image as the authoritative image (presumably because it could be recalculated fastest) but I've halfway moved away from that because it meant converting measures all the time. Not sure if that transition was completed (and/or wise).

As a side point, I think that it would be "nice" to have the uncropping when certain modules are focused only apply to the center image (where we manipulate the shapes and crop itself) but _not_ to the "second preview" image, which would _always_ be showing the "end result". Whether or not people agree on this being "nice" is a little beside the point because it would create too many complications to realise at this point. Although if a lot of systematic improvements are made for other reasons this might at some point become feasible...

> What I have tested so far

Presumably all with multiple zoom levels? It is _especially_ important to consider 200% zoom (and higher) because those are treated completely different from < 199%; a 100% zoom image is created and scaled up when displaying (whereas 199% still "extrapolates" while recalculating a full image to the needed size).

Also needing testing; different dpi and especially ppd values. The latter is only relevant for macos/retina so we'll need help from @MStraeten to confirm if that works at all.

Have you looked at rotating (both in r&p and orientation)? Is what you are seeing expected/logical/acceptable or is this behavior something you feel needs to be fixed (with possibly somewhat significant effort) before this is ready?

> * the snapshot module - a slight issue when moving the separator and dragging it seems there is a lag

While moving the separator everything is completely smooth for me. When a recalculation is needed (because of changed zoom or position (after dragging or clicking on navigator) this happens in a blocking way (and therefore there is an attempt to defer it/reduce frequency of recalcs). I don't think I've really fundamentally changed the mechanism that works these recalcs (just reconnected them to the merged processing routine). What really _should_ happen is that the separate _dev_ and _pipe_ are being run in a thread of there own, like the other three, so that updates happen smoothly. Not sure I care enough to make that happen. I think that while the update is deferred, _zooming_ is now more smooth than it was before because it correctly uses the zoom of the available (now outdated) overlay image rather than assuming it was the same as the viewport zoom.

> I have seen some skull which seems to indicate some refresh issue

I need to do more/better replumbing for this and slideshow and, most importantly, export. I thought it worked already, but this is still WIP. So far that was low priority for me in this PR, because it may need to change anyway if I decide to do any work on the caching layer; these areas will then again need to be reconnected to deal with changes I'm considering there.

> First of all thanks for taking the lead on such hard issue. After some testing it seems to me that you are definitely on the right track

Thank you! Apart from the real bugs you mentioned, what do you feel needs to still be done to make this objectively better than 4.4.2? This is one of those 80-20 projects where the last 20% could well be larger than the first 80 (and the last 10% larger still). So I'm at the point where I have to decide if I have enough commitment to "finish" it.

TurboGit commented:
> Presumably all with multiple zoom levels? It is _especially_ important to consider 200% zoom (and higher) because those are treated completely different from < 199%; a 100% zoom image is created and scaled up when displaying (whereas 199% still "extrapolates" while recalculating a full image to the needed size).

I've done my test session at different zoom level but not sure to have used all of them for Retouch or Liquify for example. I'll redo some testing on this.

dterrahe commented:
I used 4.4.2 as shorthand for "before #15084"

TurboGit commented:
> While moving the separator everything is completely smooth for me.

For me too now. So probably a bug due to the WIP status of this PR after a long testing session. Let's forget about this for now.

TurboGit commented:
I have now tested again at different zoom level (all of them, yes even Retouch at 1600 zoom level). No issue detected.

Compared to 4.4.2:
- **better**: Panning using full quality and not low res is nice and fells better (your stays into your edit not distracted by low-res blurry image).
- **better**: No big jump when releasing the panning as it was in 4.4.2 as the low-res rendering was never aligned with the actual image.
- **ok**: You don't have the big jump as discussed above but when panning you have areas on the border being temporary off and rendered as needed to replace the preview. This is OK as to me when panning you concentrate on the center area moving slightly to adjust the view for the Retouch or Liquify module. When doing the same for assessing the actual image quality (sharpness for example) again you concentrate on the center area. If moving fast or using the navigation module it can be distracting, but actually to me, with a fast computer this is ok. I have taken time to label this as OK or worst and yes I feel it is not worst just ok.
- **better**: We now have the full image displayed when using Retouch & Liquify
- **worst**: We may experience a very little pixel shift (max 1px as I see and understand) when expanding Retouch & Liquify and if a crop is active. Well worst may feel strong here, but there was no shift on 4.4.2 and now we have. But this is no big deal to me.
- **better**: If you compare to current master where expanding Retouch or Liquify on a cropped image is doing a big jump this PR is really better and this for sure lesser the point just above. 

If I have not missed something I can say that this PR is really bringing huge value.

Let me know if there is anything I have missed and should be tested.

jenshannoschwalm commented:
I spent some time testing this, especially with some "heavy masking stuff" using rasters and details and couldn't find anything not working fine yet :-) Also the user interface sems to be more fluent and visually stable. Very nice indeed.

So far on the bad side; indeed thumbnail generation and exporting is not stable yet, as mentioned above with the skulls.

dterrahe commented:
> If I have not missed something

You don't mention rotation? ;-)

Also, have you noticed the thing I mention in #15450, where after zooming in and then moving around you keep using the previous image for a bit (at "high" resolution) but then suddenly might get dropped to low resolution if you've moved away from the area where you first zoomed in (and that was calculating in the background and now appears but might be out of sight). On your fast machine you'll have to turn on some pretty heavy modules to notice this, since a zoomed image (>=200%) always calculates faster since we need fewer "pixels".

In one way it is not worse than 4.4.2 because there you'd have been in low resolution the whole time until you stop moving around. In another way, I _do_ find it somewhat disconcerting because seeing the full-res image gives the impression calculation was done and then falling back, even for the center of the image, is disorientating.

dterrahe commented:
> So far on the bad side

I _know_. :-)

That's why this is marked WIP. Slideshow is not connected up at all yet. I thought export should work, but it doesn't and needs fixing. I've not put high priority on that yet, because #15450 might cause a need further changes and doing the replumbing all over again (but maybe even simplifying it).

> I spent some time testing this

Thank you!

> Very nice indeed.

Thanks again!

TurboGit commented:
> You don't mention rotation? ;-)

Indeed :) I'll do some test with rotation. That being said some of my images where in portrait so already being rotated.

TurboGit commented:
Tested now rotation, orientation & flip. With Retouch and crop... All working fine on my side.

dterrahe commented:
What I mean is that when you _change_ rotation, you can temporarily get mismatched or misaligned overlays. Like this (or worse):
![image](https://github.com/darktable-org/darktable/assets/1549490/e5748acc-8778-4229-8ccd-90c482c184db)

I personally don't mind _too_ much (rotating anyway unavoidably causes the whole image to "jump"), but it is not "pretty".

MStraeten commented:
this pr interferes with 15451 resulting in a broken refresh of thumbnails in lighttable ...

TurboGit commented:
@MStraeten : Yes as noted in various comments this is a know bug in current draft.

dterrahe commented:
Fixed export and thumbnail generation. Also fixed the regressions (I believe) in slideshow. It still has issues (when resizing the window etc) but I _think_ they are not necessarily related to this PR. If I ever have a look at improving the cache, then slideshow could definitely benefit.

Added a toggle in the second window popover to separately switch on/off the white border for color assessment for the second preview window. This does not get switched off between sessions and when switching between darkroom and lighttable. The normal button for the mode (the little light bulb) and ctrl+b shoftcuts now only toggle for 12646 for the main/center view.
![image](https://github.com/darktable-org/darktable/assets/1549490/ba0fd881-c403-47f7-91a6-b52478124e13)

This PR still has some refreshing issues when switching between images or between views (especially in second preview window), so not marking it ready for review yet. A drag should trigger a refresh. Please report any other, non-cosmetic, issues.

jenshannoschwalm commented:
> If I ever have a look at improving the cache, then slideshow could definitely benefit.

Let's discuss this elsewhere, i have some ideas after more thinking and probably know the code better than you do so we can find out what would be needed from your side. ok?

dterrahe commented:
> This PR still has some refreshing issues when switching between images or between views (especially in second preview window)

Fixed.

I think this is good enough to go. Whatever "regressions" I still see (some flashes when switching crop, some small movements) are really not "new" per se, just become visible now that they're the _only_ marring artifacts after major moves and multiple screen refreshes are being avoided by this PR. Would still be nice to fix, but the _real_ fixes need deeper changes, imho, so I'm not really interested in more superficial workarounds.

One thing that #15084 exposed is that the preview image and the center image seem to have a slightly different roi when there's a significant crop. You'll see that when panning an image with horizontal / vertical lines; they jump sideways a little when the temporary upscaled preview fill-in gets replaced by the full resolution recalculation. Maybe this is related to the small sideways adjustment you can see when switching on/off crop (when zoomed in). I assume it is some rounding somewhere, and maybe unavoidable, but it looks to me like it can be multiple pixels so I'm not sure where that would be happening.

TurboGit commented:
Found an issue:
- Open an image in darkroom
- Right click on the second window and check or uncheck/check the color assessment setting
- Now click on the second window icon to open it

I get this on screen:

![image](https://github.com/darktable-org/darktable/assets/467069/7089a600-a88f-42bc-be13-ae345e9056f1)

And this on the console:

```
(darktable:428034): Gtk-WARNING **: 18:55:50.770: drawing failure for widget 'GtkDrawingArea': invalid matrix (not invertible)

(darktable:428034): Gtk-WARNING **: 18:55:50.770: drawing failure for widget 'GtkGrid': invalid matrix (not invertible)

(darktable:428034): Gtk-WARNING **: 18:55:50.770: drawing failure for widget 'GtkBox': invalid matrix (not invertible)

(darktable:428034): Gtk-WARNING **: 18:55:50.770: drawing failure for widget 'GtkWindow': invalid matrix (not invertible)

(darktable:428034): Gtk-WARNING **: 18:55:50.855: drawing failure for widget 'GtkDrawingArea': invalid matrix (not invertible)

(darktable:428034): Gtk-WARNING **: 18:55:50.855: drawing failure for widget 'GtkGrid': invalid matrix (not invertible)

(darktable:428034): Gtk-WARNING **: 18:55:50.855: drawing failure for widget 'GtkBox': invalid matrix (not invertible)

(darktable:428034): Gtk-WARNING **: 18:55:50.855: drawing failure for widget 'GtkWindow': invalid matrix (not invertible)

(darktable:428034): Gtk-WARNING **: 18:57:25.913: drawing failure for widget 'GtkDrawingArea': invalid matrix (not invertible)

(darktable:428034): Gtk-WARNING **: 18:57:25.913: drawing failure for widget 'GtkGrid': invalid matrix (not invertible)

(darktable:428034): Gtk-WARNING **: 18:57:25.913: drawing failure for widget 'GtkBox': invalid matrix (not invertible)

(darktable:428034): Gtk-WARNING **: 18:57:25.913: drawing failure for widget 'GtkWindow': invalid matrix (not invertible)

(darktable:428034): Gtk-WARNING **: 18:57:26.615: drawing failure for widget 'GtkDrawingArea': invalid matrix (not invertible)

(darktable:428034): Gtk-WARNING **: 18:57:26.615: drawing failure for widget 'GtkGrid': invalid matrix (not invertible)

(darktable:428034): Gtk-WARNING **: 18:57:26.615: drawing failure for widget 'GtkBox': invalid matrix (not invertible)

```

dterrahe commented:
Fixed (I think) with a little more cleanup. I had intended to restructure `dt_dev_get_zoom_scale` a little more but then forgot.

As the warnings indicate, the second window wrapped a drawingarea in a grid wrapped in a box. I can't really think why so removed those two intermediate layers and everything works fine for me. Let me know if you see any problems.

Of course the check box for the 2nd window color assessment can be assigned a shortcut and in fact I've allocated alt+b to it. Are you OK with that or would you prefer it removed? My thinking is that assigning a shortcut to a widget in one of the popovers can be done in mapping mode, if you assign a shortcut to the shortcut mapping mode and switch it on while the popover is open. But that may not be as obvious to everyone, so pre-assigning it for them (so they don't need to go into preferences) is convenient. But on the other hand, I imagine this is maybe less frequently used, so not _needed_ (it kind of makes sense to always have 2nd preview on color assessment and fit-to-screen, now that it doesn't impact the center screen).

TurboGit commented:
> OK with that or would you prefer it removed? 

I'll prefer removing the shortcut `Alt+b` because the second window is not used a lot and so don't think it is a good idea to take a shortcut by default.

> But on the other hand, I imagine this is maybe less frequently used

Exactly my thinking too.

Apart from that it works fine on my side now. So ready to be merged as soon as the shortcut is removed.

dterrahe commented:
On the other hand, (permanently) removing a shortcut is very easy these days and if the user already had alt+b set to something else, that will continue to take precedence. "Taking a shortcut" is not really _that_ big a deal.

So the argument to keep the shortcut is that in master it is very easy to toggle 2nd window color assessment by shortcut and this convenience would be lost. It's easier to write in release notes "shortcut is now alt+b" rather than "go figure out how to set up shortcuts in preferences and then open a GitHub account just to file a complaint that in this particular instance that's not as easy as you would like it to be because visual mapping doesn't work out of the box".

TurboGit commented:
> and if the user already had alt+b set to something else, that will continue to take precedence.

Oh that was my concern indeed, I feared that the shortcut would be removed. So ok to me.

TurboGit commented:
@dterrahe : Yet... This PR needs a conflict resolution now.

dterrahe commented:
This is what the user would see (in the hypothetical case that they had assigned both ctrl+b and alt+b to "normal" color assessment) ![image](https://github.com/darktable-org/darktable/assets/1549490/b674f123-fb9d-4a97-b3ce-d7258cf03357)

dterrahe commented:
Rebased

ralfbrown commented:
Has anyone else noticed a brief flash to gray when applying a style with this PR applied?  I just updated my installation and played around a bit, and discovered a flash to gray for perhaps 200ms followed by re-display of the image followed by the updated image once the recomputation finishes whenever I apply a style (either bound to a shortcut key or from the Styles button at bottom left in darkroom view).  Applying a preset, even one bound to a shortcut, doesn't cause this flashing.

Other than that and the already noted minor shifting of the newly-exposed area when panning, things look great!

dterrahe commented:
Thanks for your report and feedback. Really much appreciated!

There's a follow up PR to this, #15546, that smooths out some further issues. Using that, I don't see any flashing when applying a style. If you have time (...!) could you give that one a go too and report there? Or somebody else could report whether it improves this aspect for them?

There's some dragging behavior changes/"improvements" too that could use testing/feedback.
